homeassistant/components/switch/netio.py

```diff
@@ -0,0 +1,199 @@
+      4: "Lampe Droite"
+  - platform: netio
+    host: 192.168.1.43
+    port: 1234
+    username: user
+    password: pwd
+    ports:
+      1: "Nothing..."
+      4: "Lampe du fer"
+```
+
+To get pushed updates from the netio devices, one can add this lua code in the
+device interface as an action triggered on "Netio" "System variables updated"
+with an 'Always' schedule:
+
+``
+-- this will send socket and consumption status updates via CGI
+-- to given address. Associate with 'System variables update' event
+-- to get consumption updates when they show up
+
+local address='ha:8123'
+local path = '/api/netio'
+
+
+local output = {}
+for i = 1, 4 do for _, what in pairs({'state', 'consumption',
+                        'cumulatedConsumption', 'consumptionStart'}) do
+    local varname = string.format('output%d_%s', i, what)
+    table.insert(output,
+        varname..'='..tostring(devices.system[varname]):gsub(" ","|"))
+end end
+
+local qs = table.concat(output, '&')
+local url = string.format('http://%s%s?%s', address, path, qs)
+devices.system.CustomCGI{url=url}
+```
+
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/switch.netio/
+
+"""
+import logging
+import socket
+from homeassistant import util
+from datetime import timedelta
+from pynetio import Netio
+from collections import namedtuple
+from homeassistant.const import *
+from homeassistant.helpers import validate_config
+from homeassistant.components.switch import SwitchDevice, \
+    ATTR_CURRENT_POWER_W, ATTR_TOTAL_CONSUMPTION_KWH
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['http']
+REQUIREMENTS = ['pynetio>=0.1.5.1']
+DEFAULT_USERNAME = 'admin'
+DEFAULT_PORT = 1234
+REQ_CONF = [CONF_HOST, CONF_PORTS]
+URL_API_NETIO_EP = "/api/netio"
+
+device = namedtuple('device', ['netio', 'entities'])
+DEVICES = {}
+ATTR_START_DATE = 'Start date'
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+
+def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    """ Configure the netio linl """
+
+    if validate_config({"conf": config}, {"conf": [CONF_PORTS,
+                                                   CONF_HOST]}, _LOGGER):
+        if not config[CONF_HOST] in DEVICES:
+            try:
+                dev = Netio(config[CONF_HOST],
+                            config.get(CONF_PORT, DEFAULT_PORT),
+                            config.get(CONF_USERNAME, DEFAULT_USERNAME),
+                            config.get(CONF_PASSWORD, DEFAULT_USERNAME))
+                DEVICES[config[CONF_HOST]] = device(dev, [])
+            except:
+                _LOGGER.error('Cannot connect to %s' % config[CONF_HOST])
+
+        if DEVICES.get(config[CONF_HOST]):
+            for key in config[CONF_PORTS]:
+                switch = NetioSwitch(DEVICES[config[CONF_HOST]].netio, key,
+                                     config[CONF_PORTS][key])
+                add_devices_callback([switch])
```

Comment:
The `add_devices_callback` should only be called once at the end of your `setup_platform` function. While looping you should make a list of all the devices to add and then call this once.